aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -5,75 +5,247 @@
+
 #include <ATen/core/boxing/KernelFunction.h>
 
-namespace at {
-struct Dimname;
-}
+#include <c10/util/Metaprogramming.h>
 
 namespace c10 {
 namespace impl {
 
+//
+// utils
+//
+
+// is_tensor_ref
+template<class T> struct is_tensor_ref : std::false_type {};
+template <> struct is_tensor_ref<at::Tensor&> : std::true_type {};
+
+// is_tuple_of_tensor_refs
+//
+template<class T, class Enable = void>
+struct is_tuple_of_tensor_refs : std::false_type {};
+
+template<class T>
+struct is_tuple_of_tensor_refs<T, std::enable_if_t<guts::is_instantiation_of<std::tuple, T>::value, void>>
+: guts::typelist::all<is_tensor_ref, guts::typelist::from_tuple_t<T>>
+{};
+
+// has_ivalue_to
+//
+template<class T, class Enable = void>
+struct has_ivalue_to : std::false_type {};
+
+template<class T>
+struct has_ivalue_to<T, guts::void_t<decltype(std::declval<IValue>().to<T>())>>
+: std::true_type
+{};
+
+//
+// boxing predicates
+//
+
+// A boxable arg type is one that IValue has a constructor for.
 // Assume T is decayed
 template <typename T>
-using not_ok_to_box = guts::negation<guts::disjunction<
+using can_box =
+  guts::disjunction<
     std::is_constructible<IValue, T>,
     // TensorOptions are not directly constructible into IValue,
     // but torch::jit::push knows how to handle them
-    std::is_same<TensorOptions, T>,
-    // void returns are ok
-    std::is_same<void, T>>>;
+    std::is_same<TensorOptions, T>
+  >;
 
-// TODO boxing should be ok for all kernels. Then remove not_ok_to_box and supports_boxing.
+template <typename... Ts>
+using can_box_all = guts::conjunction<can_box<std::decay_t<Ts>>...>;
 
-template <class Result, class... Args>
-using supports_boxing =
-  guts::negation<guts::disjunction<
-    std::is_lvalue_reference<Result>,
-    not_ok_to_box<Result>,
-    std::is_same<IntArrayRef, Result>,
-    not_ok_to_box<std::decay_t<Args>>...
-  >>;
+// an unboxable result is one that can be extracted from an IValue
+template <typename T>
+using can_unbox =
+  guts::conjunction<
+    guts::disjunction<
+      has_ivalue_to<T>,
+      // void returns are ok
+      std::is_same<void, T>
+    >,
+    guts::negation<std::is_lvalue_reference<T>>
+  >;
+
+//
+// BoxedKernelWrapper
+//
+// For a given function type FT, BoxedKernelWrapper<FT> implements
+// a `call` method that
+// - takes a boxed kernel and unboxed arguments as specified by FT
+// - boxes the arguments
+// - calls the boxed kernel
+// - unboxes and returns the result
+//
+// The partial specializations below handle various cases: in
+// particular, not all types appearing in op signatures are supported,
+// and ops returning references have nonstandard wrapper implementations.
+//
+
+// base definition should never be instantiated.
+// A "no call method defined on BoxedKernelWrapper" compile error means that
+// an op signature has failed to trigger any of the partial specializations
+// that follow.
+//
+template<class FuncType, class Enable = void>
+struct BoxedKernelWrapper {};
```

Comment:
pro-tip: You could also put a static assert like `static_assert(sizeof(FuncType) == -1, "msg")` to customize the error message when partial specialization fails